libbeat/processors/decompress/gzip.go

```diff
@@ -0,0 +1,153 @@
+// specific language governing permissions and limitations
+// under the License.
+
+package gzip
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/libbeat/beat"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/processors"
+	"github.com/elastic/beats/libbeat/processors/checks"
+
+	"bytes"
+	"compress/gzip"
+	"io"
+)
+
+type decompressGzipFields struct {
+	decompressGzipFieldsConfig
+	fields map[string]string
+}
+
+type decompressGzipFieldsConfig struct {
+	Fields        common.MapStr `config:"fields"`
+	IgnoreMissing bool          `config:"ignore_missing"`
+	OverwriteKeys bool          `config:"overwrite_keys"`
+	FailOnError   bool          `config:"fail_on_error"`
+}
+
+var (
+	defaultDecompressGzipFieldsConfig = decompressGzipFieldsConfig{
+		FailOnError: true,
+	}
+)
+
+func init() {
+	processors.RegisterPlugin("decompress_gzip_fields",
+		checks.ConfigChecked(NewDecompressGzipFields,
+			checks.RequireFields("fields"),
+			checks.AllowedFields("fields", "ignore_missing", "overwrite_keys", "overwrite_keys", "fail_on_error")))
+}
+
+// NewDecompressGzipFields construct a new decompress_gzip_fields processor.
+func NewDecompressGzipFields(c *common.Config) (processors.Processor, error) {
+	config := defaultDecompressGzipFieldsConfig
+
+	err := c.Unpack(&config)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unpack the decompress_gzip_fields configuration: %+v", err)
+	}
+	if len(config.Fields) == 0 {
+		return nil, errors.New("no fields to decompress configured")
+	}
+	f := &decompressGzipFields{decompressGzipFieldsConfig: config}
+
+	// Set fields as string -> string
+	f.fields = make(map[string]string, len(config.Fields))
+	for src, dstIf := range config.Fields.Flatten() {
+		dst, ok := dstIf.(string)
+		if !ok {
+			return nil, errors.Errorf("bad destination mapping for %s: destination field must be string, not %T (got %v)", src, dstIf, dstIf)
+		}
+		f.fields[src] = dst
+	}
+	return f, nil
+}
+
+// Run applies the decompress_gzip_fields processor to an event.
+func (f *decompressGzipFields) Run(event *beat.Event) (*beat.Event, error) {
+	saved := *event
+	if f.FailOnError {
+		saved.Fields = event.Fields.Clone()
+		saved.Meta = event.Meta.Clone()
+	}
+	for src, dest := range f.fields {
+		if err := f.decompressGzipField(src, dest, event); err != nil && f.FailOnError {
+			return &saved, err
+		}
+	}
+	return event, nil
+}
+
+func (f *decompressGzipFields) decompressGzipField(src, dest string, event *beat.Event) error {
+	// Check source value
+	data, err := event.GetValue(src)
+	if err != nil {
+		if f.IgnoreMissing && errors.Cause(err) == common.ErrKeyNotFound {
+			return nil
+		}
+		return errors.Wrapf(err, "could not fetch value for field %s", src)
+	}
+
+	text, ok := data.(string)
+	if !ok {
+		return errors.Errorf("field %s is not of string type", src)
+	}
+
+	//Uncompress gzip message
+	b := bytes.NewBuffer([]byte(text))
```

Comment:
To eliminate converting to `[]byte`, you could call `bytes.NewBufferString` constructor of `bytes.Buffer`.